python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -156,1253 +120,308 @@ class MakeBrighterFatterKernelTaskConfig(pexConfig.Config):
-            means[ampName] = [[value*gain for value in pair] for pair in np.array(means[ampName])[mask]]
-            xcorrs[ampName] = [arr*gain*gain for arr in np.array(xcorrs[ampName])[mask]]
-        return
-
-    def _makeCroppedExposures(self, exp, gains, level):
-        """Prepare exposure for cross-correlation calculation.
-
-        For each amp, crop by the border amount, specified by
-        config.nPixBorderXCorr, then rescale by the gain
-        and subtract the sigma-clipped mean.
-        If the level is 'DETECTOR' then this is done
-        to the whole image so that it can be cross-correlated, with a copy
-        being returned.
-        If the level is 'AMP' then this is done per-amplifier,
-        and a copy of each prepared amp-image returned.
+                scaledCorrList.append(scaled)
+                self.log.info("Amp: %s %d/%d  Final: %g  XcorrCheck: %f",
+                              ampName, xcorrNum, len(xCorrList), q[0][0], xcorrCheck)
 
-        Parameters:
-        -----------
-        exp : `lsst.afw.image.exposure.ExposureF`
-            The exposure to prepare
-        gains : `lsst.cp.pipe.makeBrighterFatterKernel.BrighterFatterGain`
-            The object holding the amplifier gains, essentially a
-            dictionary of the amplifier gain values, keyed by amplifier name
-        level : `str`
-            Either `AMP` or `DETECTOR`
+            if len(scaledCorrList) == 0:
+                self.log.warn("Amp: %s All inputs rejected for amp!", ampName)
+                bfk.meanXcorrs[ampName] = np.zeros(bfk.shape)
+                bfk.ampKernels[ampName] = np.zeros(bfk.shape)
+                bfk.valid[ampName] = False
+                continue
 
-        Returns:
-        --------
-        scaledMaskedIms : `dict` [`str`, `lsst.afw.image.maskedImage.MaskedImageF`]
-            Depending on level, this is either one item, or n_amp items,
-            keyed by detectorId or ampName
-
-        Notes:
-        ------
-        This function is controlled by the following config parameters:
-        nPixBorderXCorr : `int`
-            The number of border pixels to exclude
-        nSigmaClipXCorr : `float`
-            The number of sigma to be clipped to
-        """
-        assert(isinstance(exp, afwImage.ExposureF))
+            if self.config.useAmatrix:
+                # This is mildly wasteful
+                preKernel = np.pad(self._tileArray(np.array(inputPtc.aMatrix[ampName])), ((1, 1)))
+            elif self.config.correlationQuadraticFit:
+                # Use a quadratic fit to the correlations as a function of flux.
+                preKernel = self.quadraticCorrelations(scaledCorrList, fluxes, f"Amp: {ampName}")
+            else:
+                # Use a simple average of the measured correlations.
+                preKernel = self.averageCorrelations(scaledCorrList, f"Amp: {ampName}")
```

Comment:
I assume that these are options to initialize the kernel and then solve for the final kernel later. Perhaps you could add a comment? Also, in the comments, you seem to imply that the first (`a` matrices) and the second options (`quadratic`) are redundant. Does this imply that, in future work, we would need to determine if we need them or not? For example, if they provide better convergence when calculating the kernel via SOR? I guess my question is, why do we have these two extra options if you say that they are mildly wasteful?

Reply:
I'll clarify that comment.  The `useAmatrix` branch still does all of the amp-wise analysis, but then doesn't use it in the final fit (although it's still persisted in the calibration).  The quadratic solver seems very sensitive to outliers when I tested it, but I didn't want to remove existing functionality in a gen2-gen3 conversion.